packages/expo-file-system/ios/EXFileSystem/EXSessionTasks/EXSessionTaskDispatcher.m

```diff
@@ -0,0 +1,114 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+#import <EXFileSystem/EXSessionTaskDispatcher.h>
+#import <EXFileSystem/EXSessionResumableDownloadTaskDelegate.h>
+
+@interface EXSessionTaskDispatcher ()
+
+@property (nonatomic, strong) NSMutableDictionary<NSURLSessionTask *, EXSessionTaskDelegate *> *tasks;
+@property (nonatomic, strong) NSMutableDictionary<NSString *, NSURLSessionDownloadTask *> *resumableDownloads;
+@property (nonatomic) BOOL isActive;
+
+@end
+
+@implementation EXSessionTaskDispatcher
+
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _tasks = [NSMutableDictionary dictionary];
+    _resumableDownloads = [NSMutableDictionary dictionary];
+    _isActive = true;
+  }
+  return self;
+}
+
+#pragma mark - public methods
+
+- (void)registerTaskDelegate:(EXSessionTaskDelegate *)delegate forTask:(NSURLSessionTask *)task
+{
+  _tasks[task] = delegate;
+}
+
+- (void)registerResumableDownloadTask:(NSURLSessionDownloadTask *)task uuid:(NSString *)uuid
+{
+  _resumableDownloads[uuid] = task;
+}
+
+- (NSURLSessionDownloadTask * _Nullable)resumableDownload:(NSString *)uuid
+{
+  return _resumableDownloads[uuid];
+}
+
+- (void)removeResumableDownload:(NSString *)uuid
+{
+  NSURLSessionTask *task = _resumableDownloads[uuid];
+  if (!task)  {
+    return;
+  }
+  
+  [_resumableDownloads removeObjectForKey:uuid];
+  [_tasks removeObjectForKey:task];
+}
+
+- (void)deactivate
+{
+  _isActive = false;
+}
+
+#pragma mark - helpers
+
+- (void)_unregisterIfResumableTaskDelegate:(EXSessionTaskDelegate *)taskDelegate
+{
+  if ([taskDelegate isKindOfClass:[EXSessionResumableDownloadTaskDelegate class]]) {
+    EXSessionResumableDownloadTaskDelegate *exResumableTask = (EXSessionResumableDownloadTaskDelegate *)taskDelegate;
+    [_resumableDownloads removeObjectForKey:exResumableTask.uuid];
+  }
+}
+
+#pragma mark - dispatcher
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
+{
+  if (_isActive) {
+    EXSessionTaskDelegate *exTask = _tasks[downloadTask];
+    if (exTask) {
+      [exTask URLSession:session downloadTask:downloadTask didFinishDownloadingToURL:location];
+      [_tasks removeObjectForKey:downloadTask];
+      [self _unregisterIfResumableTaskDelegate:exTask];
+    }
+  }
+}
+
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+  if (_isActive) {
```

Comment:
Yes, we deactivate the dispatcher before `invalidateAndCancel` because we don't want to act on the task events.
We are in the middle of the deallocating of the react context. So, I believe we shouldn't reject the Promise or inform of the termination. Tasks will be canceled, but we don't know when.

Reply:
Do you think that adding a separate `invalidate` call to native modules would make it easier to handle this situation? If we `invalidatedAndCancelled` sessions while invalidating modules maybe the `task:didCompleteWithError:` would be called in time before deallocating the bridge and a Promise rejection would come through? If this would be the case I think it could be worth to implement it… 🤔